src/runtime/tracing.cpp

```diff
@@ -10,7 +10,105 @@ typedef int32_t (*trace_fn)(void *, const halide_trace_event_t *);
 
 namespace Halide { namespace Runtime { namespace Internal {
 
-WEAK int halide_trace_file = 0;
+// A spinlock that allows for shared and exclusive access. It's
+// equivalent to a reader-writer lock, but in my case the "readers"
+// will actually be writing simultaneously to the trace buffer, so
+// that's a bad name. We use the __sync primitives used elsewhere in
+// the runtime for atomic work. They are well supported by clang.
+class SharedExclusiveSpinLock {
+    uint32_t lock;
+    const static uint32_t exclusive_held_mask = 0x80000000;
+    const static uint32_t exclusive_waiting_mask = 0x40000000;
+    const static uint32_t shared_mask = 0x3fffffff;
+
+public:
+    void acquire_shared() {
+        while (1) {
+            uint32_t x = lock & shared_mask;
+            if (__sync_bool_compare_and_swap(&lock, x, x + 1)) {
+                return;
+            }
+        }
+    }
+
+    void release_shared() {
+        __sync_fetch_and_sub(&lock, 1);
+    }
+
+    void acquire_exclusive() {
+        while (1) {
+            __sync_fetch_and_or(&lock, exclusive_waiting_mask);
+            if (__sync_bool_compare_and_swap(&lock, exclusive_waiting_mask, exclusive_held_mask)) {
+                return;
+            }
+        }
+    }
+
+    void release_exclusive() {
+        __sync_fetch_and_xor(&lock, exclusive_held_mask);
+    }
+
+    SharedExclusiveSpinLock() : lock(0) {}
+};
+
+class TraceBuffer {
+    SharedExclusiveSpinLock lock;
+    uint32_t cursor;
+    uint8_t buf[1024*1024];
+
+public:
+
+    // Attempt to atomically acquire space in the buffer to write a
+    // packet. Returns NULL if the buffer was full.
+    halide_trace_packet_t *try_acquire_packet(uint32_t size) {
+        lock.acquire_shared();
+        uint32_t my_cursor = __sync_fetch_and_add(&cursor, size);
+        if (my_cursor + size > sizeof(buf)) {
+            __sync_fetch_and_sub(&cursor, size);
+            lock.release_shared();
+            return NULL;
+        } else {
+            return (halide_trace_packet_t *)(buf + my_cursor);
+        }
+    }
+
+    // Wait for all writers to finish with their packets, stall any
+    // new writers, and flush the buffer to the fd.
+    void flush(void *user_context, int fd) {
+        lock.acquire_exclusive();
```

Comment:
I'd consider moving the locking into acquire_packet. This would put then entire lock flow into one routine and sets things up to allow lock type escalation (shared -> exclusive directly) if there is an advantage to doing so. (At present I think the code is simpler without lock escalation.) I'm not 100% certain on this so if you feel making flush standalone with respect to locking, I can see that. I do think having all the acquire and releases in acquire_packet and release_packet will make it easier to reason about. The flipside is one is tempted to add asserts that the locks are held in the inner routines, which I've never liked. (In this case, it isn't an interface called from multiple places so it doesn't matter much.)